### PR TITLE
BUG: Index(categorical, dtype=object) not returning object dtype

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

In master we get a `CategoricalIndex` back.

Fixing this turns out to be pre-requisite to fixing a bug in `CategoricalDtype._from_values_or_dtype`, which in turn is needed to simplify the CategoricalIndex constructors (xref #32141)

### CODE REVIEW

File: pandas/tests/indexes/test_base.py [hunk 1]
```diff
@@ -387,6 +387,12 @@ def test_constructor_dtypes_to_object(self, cast_index, vals):
         assert isinstance(index, Index)
         assert index.dtype == object
 
+    def test_constructor_categorical_to_object(self):
+        # Categorical data and dtype=object should return object-dtype
```
Comment (gfyoung) [inline pandas/tests/indexes/test_base.py]: Reference PR number in the comment.

**Comment by jreback** [comment]
lgtm. rebase and ping on green.

**Comment by jreback** [comment]
let's rebase this (as merged other 2 PRs)

**Comment by jbrockmendel** [comment]
rebased+green

**Comment by jreback** [comment]
thanks

**Pull request merged**